contrib/slurm/slurmexecutor.py

```diff
@@ -164,15 +164,27 @@ def execute(self, run):
         logging.debug("Command line of run is %s", args)
 
         try:
-            with open(run.log_file, "w") as f:
-                for _ in range(6):
-                    f.write(os.linesep)
-
-            run_result = run_slurm(
-                self.benchmark,
-                args,
-                run.log_file,
-            )
+            attempts = 0
+            while True:
+                run_result = run_slurm(
+                    self.benchmark,
+                    args,
+                    run.log_file,
+                )
+                if (
+                    "terminationreason" not in run_result
+                    or not run_result["terminationreason"] == "killed"
+                    or (attempts >= self.benchmark.config.retry >= 0)
+                    or STOPPED_BY_INTERRUPT
+                ):
+                    break
+                attempts += 1
+                time.sleep(1)  # as to not overcrowd a failing scheduler
```

Comment:
If this is relevant, maybe a global rate limiting across all threads is needed?

Reply:
I don't think I can justify ratelimiting globally - this is very dependent on other jobs in the queue and their workloads, so this code would only get invoked if there is a problem with the scheduler. I can safely spawn 2000 jobs in parallel some times, but fail to spawn even 20 in other cases - and in the latter case, I don't want to overload the system with fast, repeated attempts. However, if you think this needs to be addressed, I'm open to suggestions on how we should achieve that.

Reply:
I guess you would need some adaptive rate limit to handle this properly - submit jobs in parallel until something goes wrong, then reduce the parallelism. But I don't know how to implement this easily in Python and whether it is worth it. I was just wondering whether slowing down a single thread makes sense when the other parallel worker threads will still "hammer" SLURM as fast as they can.

Reply:
Only if they haven't experienced the same issue -- during most of the execution of each thread, the worker nodes work on the verification task. During that time, we don't submit new jobs from the same thread, so unless there is only a single thread failing this way (unlikely), all of the threads will be slowed down in this degraded state. While not a proper and fully-fledged solution, I think it's worth keeping in here for cases when multiple threads/processes failed to call `srun` successfully: if a 1000 parallel jobs keep retrying to no avail, I'd rather see them wait a second between each action rather than go as fast as they can.

Reply:
There is such an interface (`sbatch`), but I am mostly unfamiliar with it. And so far, even with 1000s of parallel jobs, I found no serious resource problems with the current solution.